File: server/src/main/java/de/hsrm/mi/swtpro03/server/network/NetworkService.java
package de.hsrm.mi.swtpro03.server.network;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import javax.jms.Connection;
import javax.jms.Destination;
import javax.jms.JMSException;

import org.apache.activemq.ActiveMQConnectionFactory;
import org.apache.activemq.broker.BrokerService;

/**
 * This class provides static access to all networking features
 * @author lvonk001, tschn002
**/
public class NetworkService {

	// ServerSettings
	public static String CONNECTOR_IP = "tcp://0.0.0.0:";
	public static String CONNECTION_URL = "vm://localhost";
	public static String CONNECTION_MANAGER_TOPIC = "Server";
	public static long CONNECTION_MANAGER_TIMOUT = 6000;

	private static BrokerService broker = null;
	private static Connection connection = null;

	private static ConnectionManager connectionManager;
	private static Map<String, ComChannel> allComChannels;

	
	/**
	 * Creates and starts an active MQ message broker, establishes a local connection, 
	 * adds a TCP-connector with the designated port and starts connection managment, 
	 * which listens for incoming connection event messages, such as "Connect", 
	 * "Disconnect", and "Keepalive".
	 * @param port The port for the brokers TCP-connector
	 */
	public static void start(int port) {
		if (broker != null) {
			return;
		}

		try {
			broker = new BrokerService();
			broker.setPersistent(false);

			broker.addConnector(CONNECTOR_IP + port);
			broker.start();

			ActiveMQConnectionFactory factory = new ActiveMQConnectionFactory(CONNECTION_URL);
			connection = factory.createConnection();
			connection.start();

			connectionManager = new ConnectionManager(connection, CONNECTION_MANAGER_TOPIC);
			allComChannels = Collections.synchronizedMap(new HashMap<>());

		} catch (Exception e) {
			System.err.println("[NetworkService] failed to initialize");
			e.printStackTrace();
			shutdown();
		}
	}

	/**
	 * Shuts down the message broker and closes all network resources
	 */
	public static void shutdown() {
		if (allComChannels != null) {
			for (ComChannel channel : allComChannels.values()) {
				channel._closeInternal();
			}
			allComChannels = null;
		}

		try {
			if (connectionManager != null) {
				connectionManager.close();
				connectionManager = null;
			}
			if (connection != null) {
				connection.close();
				connection = null;
			}
			if (broker != null) {
				broker.stop();
				broker = null;
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	/**
	 * Add a connection listener. 
	 * All connection listeners get notified, when a Client connects, 
	 * disconnects or fails to send keepalive messages.
	 * @param listener A connection listener
	 */
	public static void addConnectionListener(IConnectionListener listener) {
		if (connectionManager == null) {
			System.err.println("[NetworkService] not initialized");
			return;
		}
		connectionManager.addConnectionListener(listener);
	}

	/**
	 * Creates a ComChannel with the specified channel name. start() has to be called first.
	 * @param channelID The name/ID of the ComChannel
	 * @return The created ComChannel or null if something went wrong
	 */
	public static ComChannel createComChannel(String channelID) {
		if (connection == null) {
			System.err.println("[NetworkService] not initialized");
			return null;
		}
		if (allComChannels.containsKey(channelID)) {
			System.err.println("[NetworkService] duplicate channelID: " + channelID);
			return null;
		}

		ComChannel channel = new ComChannel(connection, channelID);
		allComChannels.put(channelID, channel);

		return channel;
	}

	/**
	 * Used internally. Removes a ComChannel from the ComChannel managment, after calling ComChannel.close().
	 * @param channelID The name/ID of the ComChannel
	 */
	protected static void _removeChannel(String channelID) {
		allComChannels.remove(channelID);
	}

	/**
	 * Used internally by ConnectionManager. Creates a client object with an active MQ tempQueue.
	 * @param clientQueueDestination The clients tempQueue destination
	 * @return The created client or null if something went wrong
	 */
	protected static Client _createClient(Destination clientQueueDestination) {
		if (connectionManager == null) {
			System.err.println("[NetworkService] not initialized");
			return null;
		}

		return new Client(connection, clientQueueDestination);
	}

	/**
	 * Used internally by ComChannel. Sends a direct 1-1 message to the specified client using its tempQueue.
	 * If the clientID is invalid, the call is ignored.
	 * @param channelID The name/ID of the source channel
	 * @param clientID The ID of the target client
	 * @param messageType The type of the message
	 * @param message The actual message
	 */
	protected static void _sendToClient(String channelID, String clientID, String messageType, String message) {
		Client client = connectionManager.getClientChannel(clientID);
		if (client != null)
			client.sendMessage(channelID, messageType, message);
	}

}

File: client/Client/Network/NetworkService.cs
using Apache.NMS;
using Apache.NMS.ActiveMQ;
using Apache.NMS.ActiveMQ.Commands;
using System;
using System.Collections.Generic;
using System.Threading;

namespace LoadRunnerClient.Network
{
	/// <summary>
	/// This class provides static access to all networking features
	/// </summary>
    public class NetworkService
    {
        public static string CONNECTION_TOPIC = "Server";
        public static int KEEPALIVE_MESSAGE_INTERVAL = 1000;

        private const string CONNECTION_EVENT_PROPERTY = "ConnectionEvent";
        private const string CONNECT = "Connect";
        private const string DISCONNECT = "Disconnect";
        private const string KEEPALIVE = "Keepalive";

        private static IConnection connection = null;
        private static ISession mainSession = null;

        private static ITemporaryQueue localTempQueue = null;
        private static IMessageConsumer serverMessageConsumer = null;

        private static ITopic connectionTopic = null;
        private static IMessageProducer connectionMessageProducer = null;

        private static Dictionary<string, ComChannel> allComChannels = null;

		/// <summary>
		/// The name of the clients automatically generated active MQ tempQueue servers as a unique clientID
		/// </summary>
        public static string ClientID
        {
            get; private set;
        }

		/// <summary>
		/// True if the NetworkService is connected to a server
		/// </summary>
		public static bool IsConnected 
		{
			get {
				if (connection == null)
					return false;
				if (!connection.IsStarted)
					return false;
				return true;
			}
		}

		/// <summary>
		/// Connect to a server and start sending Keepalive messages
		/// on an internally used connection event topic.
		/// </summary>
		/// <param name="url">The server adress</param>
		public static void Connect(string url) {
            if (IsConnected) {
                return;
            }

            IConnectionFactory connectionFactory = new ConnectionFactory(url);
            connection = connectionFactory.CreateConnection();
            connection.Start();

            mainSession = connection.CreateSession(AcknowledgementMode.AutoAcknowledge);

            localTempQueue = mainSession.CreateTemporaryQueue();
            serverMessageConsumer = mainSession.CreateConsumer(localTempQueue);
            serverMessageConsumer.Listener += DispatchMessageToChannel;
            ClientID = localTempQueue.ToString();

            connectionTopic = new ActiveMQTopic(CONNECTION_TOPIC);
            connectionMessageProducer = mainSession.CreateProducer(connectionTopic);

            allComChannels = new Dictionary<string, ComChannel>();

            SendConnectionEventMessage(CONNECT);
            StartKeepaliveMessageThread();

            if (!IsConnected)
            {
                Console.Error.WriteLine("[NetworkService] failed to connect.");
                Close();
            }
        }

		/// <summary>
		/// Creates a ComChannel with the specified channel name/ID. Connect() has to be called first.
		/// </summary>
		/// <param name="channelID">The name/ID of the ComChannel</param>
		/// <returns></returns>
		public static ComChannel CreateComChannel(string channelID)
        {
            if (!IsConnected)
            {
                Console.Error.WriteLine("[NetworkService] not connected");
                return null;
            }

			lock (allComChannels) {
				if (allComChannels.ContainsKey(channelID)) {
					Console.Error.WriteLine("[NetworkService] duplicate channelID: " + channelID);
					return null;
				}

				ComChannel comChannel = new ComChannel(connection, channelID);
				allComChannels.Add(channelID, comChannel);

				return comChannel;
			}			
        }

		/// <summary>
		/// Sends a disconnect event message to the server, disconnects and closes all network resources.
		/// </summary>
        public static void Disconnect()
        {			
            if (connectionMessageProducer != null)
                SendConnectionEventMessage(DISCONNECT);
            Close();
        }

		/// <summary>
		/// Disconnect and close all network resources without sending a disconnect event message.
		/// </summary>
        public static void Close()
        {
			//lock (allComChannels) {
			//	if (allComChannels != null) {
			//		foreach (ComChannel channel in allComChannels.Values) {
			//			Console.WriteLine(channel.channelID);
			//			channel._CloseInternal();
			//		}
			//		allComChannels = null;
			//	}
			//}

			if (connectionMessageProducer != null) {
                connectionMessageProducer.Close();
                connectionMessageProducer = null;
            }        

			if (serverMessageConsumer != null) {
                serverMessageConsumer.Close();
                serverMessageConsumer = null;
            }

			if (localTempQueue != null) {
				localTempQueue.Dispose();
				localTempQueue = null;
			}

			if (connectionTopic != null) {
				connectionTopic.Dispose();
				connectionTopic = null;
			}

			if (mainSession != null) {
                mainSession.Close();
                mainSession = null;
            }

			if (connection != null) {
                connection.Close();
                connection = null;
            }		
        }

		/// <summary>
		/// Used internally. Removes a ComChannel from the ComChannel managment, after calling ComChannel.Close().
		/// </summary>
		/// <param name="channelID">The name/ID of the ComChannel</param>
		internal static void _RemoveChannel(string channelID)
        {
			lock (allComChannels) {
				if (allComChannels.ContainsKey(channelID)) {
					allComChannels.Remove(channelID);
				}
			}
        }

		/// <summary>
		/// Used internally. Direct 1-1 Messages from the server are sent via a tempQueue. 
		/// They are dispatched to their corresponding ComChannel, where they are handled just like 1-n messages.
		/// </summary>
		/// <param name="amqMessage">The received direct 1-1 message</param>
        private static void DispatchMessageToChannel(IMessage amqMessage)
        {
            string channelID = null;
            channelID = amqMessage.Properties.GetString(ComChannel.CHANNELID_PROPERTY);

            if (channelID == null)
            {
                Console.Error.WriteLine("[NetworkService] Received server message without channelID");
                return;
            }
            else if (channelID.Equals(string.Empty))
            {
                Console.Error.WriteLine("[NetworkService] Received server message without channelID");
                return;
            }

            if (!allComChannels.ContainsKey(channelID))
            {
                Console.Error.WriteLine("[NetworkService] Received server message for unknown channelID: " + channelID);
                return;
            }

            ComChannel channel = allComChannels[channelID];
            channel._OnMessage(amqMessage);
        }

		/// <summary>
		/// Starts a thread to send Keepalive messages in a loop
		/// </summary>
        private static void StartKeepaliveMessageThread()
        {
            Thread keepaliveThread = new Thread(AsyncKeepaliveMessageLoop);
            keepaliveThread.Start();
        }

		/// <summary>
		/// Sends Keepalive messsages to the server in regular intervals. 
		/// Send interval length can be set with KEEPALIVE_MESSAGE_INTERVAL in milliseconds.
		/// </summary>
		private static void AsyncKeepaliveMessageLoop()
        {
            while (true)
            {
                Thread.Sleep(KEEPALIVE_MESSAGE_INTERVAL);
                if (!IsConnected)
                    return;

                SendConnectionEventMessage(KEEPALIVE);
            }
        }

		/// <summary>
		/// Sends a connection event message on a designated internal message topic.
		/// </summary>
		/// <param name="eventType">The type of the connection event</param>
        private static void SendConnectionEventMessage(String eventType)
        {
            IMessage amqMessage = mainSession.CreateMessage();
            amqMessage.NMSReplyTo = localTempQueue;
            amqMessage.Properties.SetString(CONNECTION_EVENT_PROPERTY, eventType);
            connectionMessageProducer.Send(amqMessage);
        }
    }
}